File: model/src/main/java/com/itransition/fanfictionbackend/model/Fanfic.java
package com.itransition.fanfictionbackend.model;

import lombok.Data;
import lombok.EqualsAndHashCode;

import java.util.ArrayList;
import java.util.List;
import javax.persistence.Entity;
import javax.persistence.JoinColumn;
import javax.persistence.JoinTable;
import javax.persistence.ManyToMany;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.Table;

@Data
@Entity
@Table(name = "fanfic")
@EqualsAndHashCode(callSuper = true)
public class Fanfic extends WithLongId {

    private String title;
    private String description;

    @ManyToOne(optional = false)
    @JoinColumn(name = "author_id")
    private User author;

    @OneToMany(mappedBy = "fanfic", orphanRemoval = true)
    private List<FanficChapter> chapters = new ArrayList<>();

    @OneToMany(mappedBy = "fanfic", orphanRemoval = true)
    private List<Rating> ratings = new ArrayList<>();

    @OneToMany(mappedBy = "fanfic", orphanRemoval = true)
    private List<Reaction> reactions = new ArrayList<>();

    @OneToMany(mappedBy = "fanfic", orphanRemoval = true)
    private List<Comment> comments = new ArrayList<>();

    @ManyToMany
    @JoinTable(
        name = "fanfic_tag",
        joinColumns = @JoinColumn(name = "fanfic_id"),
        inverseJoinColumns = @JoinColumn(name = "tag_id")
    )
    private List<Tag> tags = new ArrayList<>();
}

File: src/main/kotlin/com/itransition/fanfictionbackend/model/Fanfic.kt
package com.itransition.fanfictionbackend.model

import javax.persistence.Entity
import javax.persistence.JoinColumn
import javax.persistence.JoinTable
import javax.persistence.ManyToMany
import javax.persistence.ManyToOne
import javax.persistence.OneToMany
import javax.persistence.Table

@Entity
@Table(name = "fanfic")
class Fanfic : WithLongId() {

    var title: String? = null
    var description: String? = null

    @ManyToOne(optional = false)
    @JoinColumn(name = "author_id")
    var author: User? = null

    @OneToMany(mappedBy = "fanfic", orphanRemoval = true)
    var chapters: MutableList<FanficChapter> = mutableListOf()

    @OneToMany(mappedBy = "fanfic", orphanRemoval = true)
    var ratings: List<Rating> = emptyList()

    @OneToMany(mappedBy = "fanfic", orphanRemoval = true)
    var reactions: List<Reaction> = emptyList()

    @OneToMany(mappedBy = "fanfic", orphanRemoval = true)
    var comments: List<Comment> = emptyList()

    @ManyToMany
    @JoinTable(
        name = "fanfic_tag",
        joinColumns = [JoinColumn(name = "fanfic_id")],
        inverseJoinColumns = [JoinColumn(name = "tag_id")]
    )
    var tags: List<Tag> = emptyList()
}
